cpp/src/arrow/type.h

```diff
@@ -288,31 +288,46 @@ std::shared_ptr<DataType> GetPhysicalType(const std::shared_ptr<DataType>& type)
 class ARROW_EXPORT FixedWidthType : public DataType {
  public:
   using DataType::DataType;
+  // This is only for preventing defining this class in each
+  // translation unit to avoid one-definition-rule violation.
+  ~FixedWidthType() override;
```

Comment:
Curious why it causes ODR violation.

Reply:
To be honest, I don't know much about this but...

I tried to create a small example to reproduce this case but couldn't create it... So I explain what I observed:

If we don't have this, `.o` of a C++ code that has `std::make_shared<arrow::XXXType>` such as `std::make_shared<arrow::ListType>` has the following `nm` result:

```console
$ nm --demangle src/arrow/CMakeFiles/arrow_objlib.dir/compute/kernels/scalar_cast_numeric.cc.o | grep 'vtable for arrow::FloatingPointType'
00000000 W vtable for arrow::FloatingPointType
```

`W` is a global weak symbol.

https://linux.die.net/man/1/nm

>  If lowercase, the symbol is local; if uppercase, the symbol is global (external). 

> "W"
> "w"
>
> The symbol is a weak symbol that has not been specifically tagged as a weak object symbol.

If we have this, the symbol becomes a undefined symbol:

```console
$ nm --demangle src/arrow/CMakeFiles/arrow_objlib.dir/compute/kernels/scalar_cast_numeric.cc.o | grep 'vtable for arrow::FloatingPointType'
         U vtable for arrow::FloatingPointType
```

https://linux.die.net/man/1/nm

> "U"
>
> The symbol is undefined. 

If there are multiple `W vtable for arrow::FloatingPointType` in our `.o`. linking `libarrow.so` with `-flto` is failed:

```text
[650/1111] Linking CXX shared library debug/libarrow.so.1100.0.0
FAILED: debug/libarrow.so.1100.0.0 
: && /bin/c++ -fPIC -Wno-noexcept-type -flto -fdiagnostics-color=always  -Wall -Wno-conversion -Wno-sign-conversion -Wunused-result -fno-semantic-interposition -msse4.2  -g -Werror -O0 -ggdb -shared -Wl,-soname,libarrow.so.1100 -o debug/libarrow.so.1100.0.0 ...(**/*.o)... && :
cpp/src/arrow/type.h:313: error: virtual table of type ‘struct FloatingPointType’ violates one definition rule [-Werror=odr]
  313 | class ARROW_EXPORT FloatingPointType : public NumberType {
      | 
cpp/src/arrow/type.h:313:20: note: the conflicting type defined in another translation unit
  313 | class ARROW_EXPORT FloatingPointType : public NumberType {
      |                    ^
<built-in>: note: virtual method ‘__cxa_pure_virtual’
...
```

It seems that `W` and `-flto` is a bad combination. If we change `W` to `U` by defining a not-inlined deconstructor, we don't get the ODR error.

Do you notice anything from this observation?

Reply:
I don't know why it happens. Clang doesn't complain ODR issue.
cc @pitrou, @westonpace, @lidavidm for comments.

Reply:
I understand why adding that destructor changes W to U.  You are defining a non-inline key function.  Therefore the vtable is defined in only one place.

However, I don't understand why it is an ODR error.  For example, in the [clang docs](https://lld.llvm.org/missingkeyfunction.html):

> When a class has no non-pure, non-inline, virtual functions, there is no key function, and the compiler is forced to emit the vtable in every translation unit that references the class. In this case, it is emitted in a COMDAT section, which allows the linker to eliminate all duplicate copies. This is still wasteful in terms of object file size and link time, so it’s always advisable to ensure there is at least one eligible function that can serve as the key function.

Reply:
It seems that this is only happen with GCC (GNU ld?).
I couldn't reproduce this with `CC=clang CXX=clang++ CFLAGS="-flto" CXXFLAGS="-flto" cmake ...`. (`W` symbol is still generated but there is no link error.)

Is it acceptable that we add this change only for GCC and LTO?

Reply:
I think this change is okay. And I wonder how you find this solution, it's not obvious from the error log.

Reply:
It's just a try and error...

There are some abstract classes for `arrow::DataType` family. Most of them are in error messages but `arrow::DataType` isn't shown. I noticed that `arrow::DataType` is `U` and others are `W` in `nm --demangle` result. I found that `arrow::DataType::~DataType()` is defined in `type.cc` but other abstract classes don't define their destructors. So I tried to define them.

Reply:
Yes, that is what it means.  I think "wasteful" is too strong of a word though.  I think the consequence should generally be minor enough that it isn't really a concern.